interop-scoring/main.js

```diff
@@ -0,0 +1,506 @@
+      }
+    }
+    aggregateScore += Math.floor(1000 * subtestsAllPassing / results.size);
+  }
+  return Math.floor(aggregateScore / interopScores.size) || 0;
+}
+
+// Score a set of runs (independently) on a set of tests. The runs are presumed
+// to be aligned in some way (i.e. they were all run at the same WPT SHA).
+//
+// Returns an array of scores, which is the top-level score (integer 0-1000) for
+// each corresponding input run.
+//
+// To get the top-level score for a run, each test in that run that is present
+// in |allTestsSet| is examined. Each test is scored 0-1000 based on the
+// fraction of its subtests that pass, with rounding down so that 1000 means
+// all subtests pass. Reftests score either 0 or 1000. These test scores are
+// then summed and divided by the size of |allTestsSet|, again rounding down.
+//
+// This methodology has several consequences:
+//
+//   1. Individual tests do have a heavier weight than subtests. This could be
+//   gamed, by splitting passing tests into multiple files rather than using
+//   subtests (or conversely by combining failing tests into subtests in a
+//   single file).
+//
+//   2. If |allTestsSet| is constant across runs *through time*, older runs may
+//   not have entries for tests were only added recently and will be penalized
+//   for that. This is deliberate - see the comment block later in this
+//   function for why.
+//
+//   3. We could show (on wpt.fyi) scores at both the test and category level as
+//   a percentage with one decimal point, and what a user would see would be the
+//   same numbers that go into the total score, with no hidden rounding error.
+//
+//   4. Because we round down twice, the score for a category can end up lower
+//   than if we used rational numbers.
+function scoreRuns(runs, allTestsSet) {
+  const scores = [];
+  const interopScores = new Map();
+  try {
+    for (const run of runs) {
+      // Sum of the integer 0-1000 scores for each test.
+      let score = 0;
+      lib.results.walkTests(run.tree, (path, test, results) => {
+        const testname = path + '/' + test;
+        if (!allTestsSet.has(testname)) {
+          return;
+        }
+
+        // TODO: Validate the data by checking that all statuses are recognized.
+
+        let subtestPasses = 0;
+        let subtestTotal = 1;
+
+        // Keep subtest data for every test in order to calculate interop scores.
+        // A test entry is created for each test in the first run.
+        if (!interopScores.has(testname)) {
+          interopScores.set(testname, new Map());
+        }
+        if ('subtests' in results) {
+          if (results['status'] != 'OK' && !KNOWN_TEST_STATUSES.has(testname)) {
+            throw new Error(`Unexpected non-OK status for test: ${testname}`);
+          }
+          subtestTotal = results['subtests'].length;
+          for (const subtest of results['subtests']) {
+            // Keep a number that represents how many browsers passed the subtest.
```

Comment:
Just a tad simpler language:

```suggestion
            // Keep a count of how many browsers passed the subtest.
```